### ISSUE TITLE
Add gradient checkpointing

### ISSUE THREAD

**Issue by williamFalcon** [PyTorchLightning/pytorch-lightning#1308]
Would be great to support gradient checkpointing for a whole lightningModule.

@tullie

**Comment by tullie** [comment]
Happy to do this, what's the estimated timeline for the next release?

**Comment by williamFalcon** [comment]
was aiming for this upcoming friday?

**Comment by ReshinthAdith** [comment]
Still open? Any detailed insight on where to get started?

**Comment by tullie** [comment]
@ReshinthAdith we had a small discussion on Slack about this. We concluded that all lightning could support, with respect to gradient checkpointing, was a trainer flag (e.g. gradient_checkpoint), which turns on gradient checkpointing for all nn.modules in the LightningModule. The idea would be to recursively apply gradient checkpointing to every lightning module when the flag is True. If users wanted to apply it to specific modules, they'd need to turn this flag off and apply it manually (basically how it would be done atm).

Any help with the task would be great. Are you familiar with the pytorch gradient checkpointing API? Let me know if there's anyway I can help you.
